### PR TITLE
Make window invisible until RedrawEventsCleared (#5236)

### LINKED ISSUE #5236: Make window(s) start invisible then turn on visiblity on `RedrawEventsCleared` event
## What problem does this solve or what need does it fill?

By default, winit shows white/black surface depending on system, it is visible until wgpu select rendering device and run our renderer. Instead, renderer context should be automatically visible when window appears to be visible.

## What solution would you like?

Bevy should start window with `.with_visible(false)` and then set it to `true` in `RedrawEventsCleared` section.

## What alternative(s) have you considered?

None

## Additional context

None

### PR DESCRIPTION
**Pull request by HeavyRain266**
# Objective

Fixes #5236.

## Solution

Makes bevy spawn invisible winduw and exposes `set_visible()` and `visible()` to make it visible once renderer is working to prevent white/black surface from winit being displayed menawhile.

## Changelog

- Add `bevy_window::Window::set_visible()` and `bevy_window::Window::visible()` methods for manipulating window visibility options.

### CODE REVIEW

File: crates/bevy_winit/src/lib.rs [hunk 1]
```diff
@@ -612,9 +616,15 @@ pub fn winit_runner_with(mut app: App) {
             }
             Event::RedrawEventsCleared => {
                 {
-                    let winit_config = app.world.resource::<WinitSettings>();
-                    let windows = app.world.resource::<Windows>();
+                    let mut windows = app.world.resource_mut::<Windows>();
                     let focused = windows.iter().any(|w| w.is_focused());
+                    if !windows.iter().any(|w| w.visible()) {
```
Comment (IceSentry) [inline crates/bevy_winit/src/lib.rs:621]: This doesn't seem right. I don't think this code should run every frame, but also this doesn't work if you have multiple windows. Try running the multiple_windows example to see what I mean

Comment (IceSentry) [reply crates/bevy_winit/src/lib.rs:621]: If you remove the condition both windows will appear, but if you keep this condition only one window appears at a time

**Review by IceSentry — COMMENTED** [review]
I like the idea, but I think that making the window visible should be done in a startup system that runs in PostStartup.

Adding something like this to the plugins works for me:
```rust
        .add_startup_system_to_stage(
            StartupStage::PostStartup,
            |windows: Res<Windows>, winit_windows: NonSendMut<WinitWindows>| {
                let winit_window = windows
                    .get_primary()
                    .and_then(|window| winit_windows.get_window(window.id()))
                    .expect("Failed to get window");
                winit_window.set_visible(true);
            },
        )
```

**Comment by IceSentry** [comment]
Actually, that doesn't work, since it won't work for any window that gets spawned after the startup...

**Comment by HeavyRain266** [comment]
> Actually, that doesn't work, since it won't work for any window that gets spawned after the startup...

Hmm, what should I do then? remove condition and/or move it to different event?

**Comment by IceSentry** [comment]
Yeah, removing the condition is currently the simplest solution. I'm still not a fan of it running every frame, but realistically that's probably not going to be an issue. 

Like you said, this API didn't exist before so I guess it's fine, but I can already imagine people trying to use it and be surprised that the window doesn't stay invisible.

**Comment by HeavyRain266** [comment]
> I can already imagine people trying to use it and be surprised that the window doesn't stay invisible.

Personally, I can't find any usecases for hiding game window but mayby should I add note to the docs that it's intended to be used internally and you should look into switching between headless and regular mode instead? When you hide window, renderer is still working in the background which may cause a lot of confusion/performance issue for end user 🤔.

**Comment by IceSentry** [comment]
Yeah, I agree, I can't think of a use case either, but since the API is public it's a potential footgun

**Comment by IceSentry** [comment]
I guess, one option would be to just not expose the api publicly and work directly with the winit window. Essentially always create it as invisible and in the RedrawRequested modify the winit window directly instead of going through the bevy abstraction

**Comment by alice-i-cecile** [comment]
So, this "public methods that seemingly do nothing" is a serious problem that I've seen users trip over before.

However, it's also not new to this PR. I would probably suggest:

1) Adding prominent warnings to the `Window` struct explaining that it cannot be modified after initialization.
2) Opening / searching for an issue about this, and ideally linking that in the docs above.
3) Addressing this problem in a follow-up PR.

**Comment by LoipesMas** [comment]
I think this solution is pretty "hacky". And the fact that `set_visible` actually does nothing should be __explicitly__ noted imo.

But perhaps we could add a boolean flag to `Window` struct, e.g. `initialized`, that starts as `false` and on `RedrawEventsCleared` if its `false`, then window is set to visible and the flag is set to `true`. So on the next `RedrawEventsCleared` the window is not necessarily set to visible. (This checking would still happen every frame tho)
Or am I missing a reason why this wouldn't work?

**Review by alice-i-cecile — APPROVED** [review]
Documenting this is better than not, and this is quite clear. Still makes me flinch, but I won't block on a better solution.

**Comment by Isso4820** [comment]
@HeavyRain266 I'm actually against merging this because I actually wanted a feature to be able to create a headless window. My use case is I want to write an application hidden in the background that would periodically save rendered content to an image to be used somewhere else.

I was about to create a very similar Pull Request adding `visible` field to `WindowDescriptor` to be passed down in `create_window` in `winit_windows.rs` when I found this one. 

> Personally, I can't find any usecases for hiding game window 

Well... [relevant XKCD](https://xkcd.com/1172/)

**Comment by HeavyRain266** [comment]
> @HeavyRain266 I'm actually against merging this because I actually wanted a feature to be able to create a headless window. My use case is I want to write an application hidden in the background that would periodically save rendered content to an image to be used somewhere else.
> 
> I was about to create a very similar Pull Request adding `visible` field to `WindowDescriptor` to be passed down in `create_window` in `winit_windows.rs` when I found this one. 
> 
> > Personally, I can't find any usecases for hiding game window 
> 
> Well... [relevant XKCD](https://xkcd.com/1172/)
> 
> 

You don't need window for that, bevy already have headless mode...

**Comment by HeavyRain266** [comment]
Closing because will look into upstream solution in wgpu instead. This behaviour doesn't happen if you use raw Desktop APIs, Switch's NVN or PS5 SDK instead of default renderer (using bevy_window). Wgpu seems to defer device binding for X seconds which causes winit to show it's default surface instead of user's renderer context.

**Comment by IceSentry** [comment]
@HeavyRain266 I think we should still merge the window visibility controls. That way users would be able to hide the window if they need too. This would also mean that hiding the window on startup could be handled by users without any special handling by bevy. Having a fix upstream is a nice fix too, but I think there's still some value in that api. Winit supports hiding the window so we probably should support that as well.

**Comment by HeavyRain266** [comment]
@IceSentry Feel free to re-use that code in separe PR which adds ability for hiding windows 🙂

**Pull request closed without merging**